File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.cpp
#include <string>
#include <vector>

using namespace std;

string solution(string number, int k) {
    string ans = ""; // 스택 역할
    
    for (char c : number) {
        // 1. 스택이 비어있지 않고
        // 2. 아직 제거할 횟수(k)가 남았으면
        // 3. 스택의 마지막 글자가 현재 숫자보다 작으면 제거
        while (!ans.empty() && k > 0 && ans.back() < c) {
            ans.pop_back();
            k--;
        }
        ans.push_back(c);
    }
    
    // 예외 처리: k가 남은 경우 (ex: "987", k=1)
    // 이미 내림차순 정렬이 되어 pop이 안 일어났으므로 뒤에서부터 k개를 자름
    while (k > 0) {
        ans.pop_back();
        k--;
    }
    return ans;
}

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
def solution(number, k):
    answer = []  # 스택 역할

    for c in number:
        # 1. 스택이 비어 있지 않고
        # 2. 아직 제거할 횟수 k가 남아 있고
        # 3. 스택의 마지막 숫자가 현재 숫자보다 작으면 제거
        while answer and k > 0 and answer[-1] < c:
            answer.pop()
            k -= 1

        answer.append(c)

    # 예외 처리: k가 남은 경우
    # 예: number = "987", k = 1
    # 내림차순이라 앞에서 제거되지 않았으므로 뒤에서 k개 제거
    if k > 0:
        answer = answer[:-k]

    return ''.join(answer)